lib/Valetudo.js

```diff
@@ -132,4 +132,24 @@ Valetudo.VACUUM_MODEL_PROVIDER = function() {
     }
 };
 
+Valetudo.prototype.shutdown = function() {
+    console.info("Valetudo shutdown in progress...")
+    
+    //shutdwon mqtt client
+    if(this.configuration.get("mqtt") && this.configuration.get("mqtt").enabled === true) {
+        this.mqttClient.shutdown();
+    }
+
+    //shutdown webserver
+    this.webserver.shutdown();
+    
+    //shutdown vacuum
+    this.vacuum.shutdown();
+    
+    //shutdown dummycloud
+    this.dummycloud.shutdown();
+
+    console.info("Valetudo shutdown done.")
```

Comment:
Since (some of) those operations above are asynchronous, this line might appear even if they aren't shut down yet.

The async library contains some useful functions to avoid the callback hell for such applications:
https://caolan.github.io/async/v3/docs.html#parallel

I'm still looking for nice ways to achieve the same tasks without additional libraries so if someone knows something please comment. 🙃

Reply:
Refactored the calls to a async calls.

The current solution is based on callbacks, but it also introduces the thing called 'callback hell'.

I played around with Promises and async/await during the last couple of days but wasn't successful. Especially the handling of inner callbacks (mqttclient.end) with promises gave me some headaches. Maybe nodes util.promisify can help, but I'm not sure.